go/cli/mcap/cmd/filter.go

```diff
@@ -0,0 +1,317 @@
+		if err != nil {
+			die("configuration error: %s", err)
+		}
+		var reader io.Reader
+		if len(args) == 0 {
+			stat, err := os.Stdin.Stat()
+			if err != nil {
+				die("failed to check stdin state: %s", err)
+			}
+			if stat.Mode()&os.ModeCharDevice == 0 {
+				reader = os.Stdin
+			} else {
+				die("supply a file to read")
+			}
+		} else {
+			close, newReader, err := utils.GetReader(context.Background(), args[0])
+			if err != nil {
+				die("failed to open source for reading: %s", err)
+			}
+			defer func() {
+				if closeErr := close(); closeErr != nil {
+					die("error closing read source: %s", closeErr)
+				}
+			}()
+			reader = newReader
+		}
+
+		var writer io.Writer
+		if filterOutput == "" {
+			if !utils.StdoutRedirected() {
+				die("Binary output can screw up your terminal. Supply -o or redirect to a file or pipe")
+			}
+			writer = os.Stdin
+		} else {
+			newWriter, err := os.Create(filterOutput)
+			if err != nil {
+				die("failed to open %s for writing: %s", filterOutput, err)
+			}
+			writer = newWriter
+		}
+
+		err = filter(reader, writer, filterOptions)
+		if err != nil {
+			die(error.Error(err))
+		}
+	},
+}
+
+func compileMatchers(regexStrings []string) ([]regexp.Regexp, error) {
+	var matchers []regexp.Regexp
+
+	for _, regexString := range regexStrings {
+		// auto-surround with ^$ if not specified.
+		if regexString[:1] != "^" {
+			regexString = "^" + regexString
+		}
+		if regexString[len(regexString)-1:] != "$" {
+			regexString = regexString + "$"
+		}
+		regex, err := regexp.Compile(regexString)
+		if err != nil {
+			return nil, err
+		}
+		matchers = append(matchers, *regex)
+	}
+	return matchers, nil
+}
+
+type markableSchema struct {
+	*mcap.Schema
+	written bool
+}
+type markableChannel struct {
+	*mcap.Channel
+	written bool
+}
+
+func filter(
+	r io.Reader,
+	w io.Writer,
+	opts *filterOpts,
+) error {
+	lexer, err := mcap.NewLexer(r)
+	if err != nil {
+		return err
+	}
+	mcapWriter, err := mcap.NewWriter(w, &mcap.WriterOptions{
+		Compression: opts.compressionFormat,
+	})
+	if err != nil {
+		return err
+	}
+	defer func() {
+		err = mcapWriter.Close()
+	}()
+
+	buf := make([]byte, 1024)
+	schemas := make(map[uint16]markableSchema)
+	channels := make(map[uint16]markableChannel)
+
+	for {
+		token, data, err := lexer.Next(buf)
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				return nil
+			}
+		}
+		if len(data) > len(buf) {
+			buf = data
+		}
+		switch token {
+		case mcap.TokenHeader:
+			header, err := mcap.ParseHeader(data)
+			if err != nil {
+				return err
+			}
+			if err = mcapWriter.WriteHeader(header); err != nil {
+				return err
+			}
+		case mcap.TokenSchema:
+			schema, err := mcap.ParseSchema(data)
+			if err != nil {
+				return err
+			}
+			schemas[schema.ID] = markableSchema{schema, false}
+		case mcap.TokenChannel:
+			channel, err := mcap.ParseChannel(data)
+			if err != nil {
+				return err
+			}
+			if len(opts.includeTopics) > 0 {
```

Comment:
the length check here and on 224 is unnecessary - the iteration will not run if the length is zero